### DISCUSSION TITLE
How to access methods at different object path levels sharing the same interface name?

### DISCUSSION THREAD

**Discussion by seahawk1986** [python-sdbus/python-sdbus#66] [General]
Hello,
I am currently evaluating sdbus for a hobby project and stumbled on an problem with the DBus interface of https://github.com/flensrocker/vdr-plugin-dbus2vdr/ - the structure according to d-feet looks like this (all the `/Plugins/{plugin_name}` objects share the same methods and signatures):
![d_feet_vdr_interface](https://github.com/python-sdbus/python-sdbus/assets/935408/b2e479fd-0782-4845-9a6b-e3bf48edcb35)

sdbus generates those code snippets for the interfaces:
```python
$ python -m sdbus gen-from-connection --system de.tvdr.vdr /Plugins
from __future__ import annotations

from typing import Any, Dict, List, Tuple

from sdbus import (
    DbusDeprecatedFlag,
    DbusInterfaceCommonAsync,
    DbusNoReplyFlag,
    DbusPropertyConstFlag,
    DbusPropertyEmitsChangeFlag,
    DbusPropertyEmitsInvalidationFlag,
    DbusPropertyExplicitFlag,
    DbusUnprivilegedFlag,
    dbus_method_async,
    dbus_property_async,
    dbus_signal_async,
)


class DeTvdrVdrPluginInterface(
    DbusInterfaceCommonAsync,
    interface_name="de.tvdr.vdr.plugin",
):
    @dbus_method_async(
        result_signature="a(ss)",
    )
    async def list(
        self,
    ) -> List[Tuple[str, str]]:
        raise NotImplementedError


class DeTvdrVdrPluginmanagerInterface(
    DbusInterfaceCommonAsync,
    interface_name="de.tvdr.vdr.pluginmanager",
):
    @dbus_method_async(
        result_signature="a(ss)",
    )
    async def list(
        self,
    ) -> List[Tuple[str, str]]:
        raise NotImplementedError
```
This is written to a file `plugins.py`.
```python
$ python -m sdbus gen-from-connection --system de.tvdr.vdr /Plugins/softhddevice
from __future__ import annotations

from typing import Any, Dict, List, Tuple

from sdbus import (
    DbusDeprecatedFlag,
    DbusInterfaceCommonAsync,
    DbusNoReplyFlag,
    DbusPropertyConstFlag,
    DbusPropertyEmitsChangeFlag,
    DbusPropertyEmitsInvalidationFlag,
    DbusPropertyExplicitFlag,
    DbusUnprivilegedFlag,
    dbus_method_async,
    dbus_property_async,
    dbus_signal_async,
)


class DeTvdrVdrPluginInterface(
    DbusInterfaceCommonAsync,
    interface_name="de.tvdr.vdr.plugin",
):
    @dbus_method_async(
        input_signature="ss",
        result_signature="is",
    )
    async def svdrpcommand(
        self,
        command: str,
        option: str,
    ) -> Tuple[int, str]:
        raise NotImplementedError

    @dbus_method_async(
        input_signature="ss",
        result_signature="b",
    )
    async def service(
        self,
        id: str,
        data: str,
    ) -> bool:
        raise NotImplementedError
```
This is written to a file `plugin_interface.py`.

Now I want to import those Interfaces - this causes an ValueError:
```pycon
Python 3.12.2 (main, Feb 25 2024, 16:35:05) [GCC 11.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import plugins
>>> import plugin_interface
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/alexander/projects/python3-yavdrfrontend3/yavdr_frontend/dbus_interfaces/vdr/plugin_interface.py", line 20, in <module>
    class DeTvdrVdrPluginInterface(
  File "/home/alexander/projects/python3-yavdrfrontend3/.venv-3.12/lib/python3.12/site-packages/sdbus/dbus_proxy_async_interface_base.py", line 261, in __new__
    raise ValueError(
ValueError: D-Bus interface of the name 'de.tvdr.vdr.plugin' was already created.
```

`List` on both `de.tvdr.vdr.plugin` and `de.tvdr.pluginmanager` does the same thing, so I can remove the colliding `DeTvdrVdrPluginInterface` class from `plugins.py` for my purposes, but since this was never a problem with other Python dbus libraries (python-dbus, pydbus, dasbus) I am wondering if there is a way around it (or does this interface something not allowed by the DBus specification?).

**Comment by igo95862** [comment]
Hello @seahawk1986 

Having the two different interfaces share the name is a very bad practice. For example, introspection XML under `/usr/share/dbus-1/interfaces/` folder is named by their interface names so those two interfaces will conflict.

> or does this interface something not allowed by the DBus specification

There is nothing that prohibits it directly but I want to enforce good practices.

As a workaround you can merge those two interfaces in a single class:

```python
from __future__ import annotations

from typing import Any, Dict, List, Tuple

from sdbus import (
    DbusDeprecatedFlag,
    DbusInterfaceCommonAsync,
    DbusNoReplyFlag,
    DbusPropertyConstFlag,
    DbusPropertyEmitsChangeFlag,
    DbusPropertyEmitsInvalidationFlag,
    DbusPropertyExplicitFlag,
    DbusUnprivilegedFlag,
    dbus_method_async,
    dbus_property_async,
    dbus_signal_async,
)


class DeTvdrVdrPluginInterface(
    DbusInterfaceCommonAsync,
    interface_name="de.tvdr.vdr.plugin",
):
    @dbus_method_async(
        input_signature="ss",
        result_signature="is",
    )
    async def svdrpcommand(
        self,
        command: str,
        option: str,
    ) -> Tuple[int, str]:
        raise NotImplementedError

    @dbus_method_async(
        input_signature="ss",
        result_signature="b",
    )
    async def service(
        self,
        id: str,
        data: str,
    ) -> bool:
        raise NotImplementedError

    @dbus_method_async(
        result_signature="a(ss)",
    )
    async def list(
        self,
    ) -> List[Tuple[str, str]]:
        raise NotImplementedError
```

**Reply by seahawk1986** [reply]
Thank you.

![d_feet_vdr_setup_interface](https://github.com/python-sdbus/python-sdbus/assets/935408/84fa2283-fa89-486b-9bd6-fad0b1dd56e8)


Another thing I noticed was that sdbus can generate invalid method names that are protected Python keywords, e.g. a DBus Method with the name `Del()` is transformed into `async def del(self, ...)` - while linters catch those very quickly and you can change the name and override the `method_name` in the decorator, I wonder if if would be helpful to improve the name mangling or output a warning when creating the code for the interface class.

**Reply by igo95862** [reply]
My current policy is that the output of the code generator must be inspected before running. There is even a warning in the documentation not to send it to `compile()`.

I do plan to add a command line options to override a specific interface method name so that you can have a reproducible output.
